Granule to STAC Item / Kerchunk Script

agoodm opened the pull request:
This adds a script for generating Kerchunk JSON references for an image pair granule, and from these a STAC item. 

Usage:

```
python gran_to_cat.py /path/to/gran.nc
```

There is an optional --s3_path flag if processing a granule locally, but the included URLs in both JSON files will reference their S3 Path.

jhkennedy commented on src/kerchunk/gran_to_cat.py:
```diff
@@ -0,0 +1,197 @@
+    
+        
+def geom_from_refs(refs: Dict[str, Any]):
+    def maybe_scalar(g):
```
I think this would be better called `force_scaler` or `ensure_scalar` as you're extracting a scalar out of a tuple here.

jhkennedy commented on src/kerchunk/gran_to_cat.py:
```diff
@@ -0,0 +1,197 @@
+
+@click.command()
+@click.argument("path")
+@click.option("--s3_path", type=str, default=None)
```
```suggestion
@click.option("--s3-path", type=str, default=None)
```
Typically CLI arguments are kabob-ed not snak-ed and click should handle turning `-` to `_` for the actual variable.

jhkennedy reviewed (commented):
This script doesn't run for me.

I tried regenerating the proposed STAC item using the same granule, and I get a `KeyError`:
```
python gran_to_cat.py LC08_L1GT_004010_20140206_20200912_02_T2_X_LC08_L1GT_004010_20140529_20200911_02_T2_G0120V02_P008.nc --s3_path s3://its-live-data/velocity_image_pair/landsatOLI/v02/N70W040/LC08_L1GT_004010_20140206_20200912_02_T2_X_LC08_L1GT_004010_20140529_20200911_02_T2_G0120V02_P008.nc
Traceback (most recent call last):
  File ".../_test/kerchunk/gran_to_cat.py", line 197, in <module>
    make_granule_stac_kerchunk()
  File ".../mambaforge/envs/its-kerchunk/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../mambaforge/envs/its-kerchunk/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File ".../mambaforge/envs/its-kerchunk/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../mambaforge/envs/its-kerchunk/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../_test/kerchunk/gran_to_cat.py", line 190, in make_granule_stac_kerchunk
    stac = refs_to_stac_item(refs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File ".../_test/kerchunk/gran_to_cat.py", line 168, in refs_to_stac_item
    item = xstac.xarray_to_stac(DuckDataset(ref),
                                ^^^^^^^^^^^^^^^^
  File ".../_test/kerchunk/gran_to_cat.py", line 92, in __init__
    zarray = ujson.loads(r[f"{v}/.zarray"])
                         ~^^^^^^^^^^^^^^^^
KeyError: 'img_pair_info/.zarray'
```


My environment file for this script:
```yaml
name: its-kerchunk
channels:
  - conda-forge
  - nodefaults
dependencies:
  - python
  - click
  - fsspec
  - geopandas
  - numpy
  - pandas
  - pyproj
  - shapely
  - ujson
  - xstac
  - kerchunk
  - netcdf4
  - h5py
  - ca-certificates
  - certifi
  - openssl
```
and the environment listing:
[env_list.txt](https://github.com/user-attachments/files/17139737/env_list.txt)

agoodm commented on src/kerchunk/gran_to_cat.py:
```diff
@@ -0,0 +1,197 @@
+import os
```
I prefer sticking with `os` because in this case I am just using `os.path.basename` to extract the filename (with the extension) from the granule URL. But IIRC, doing the same thing is two lines with `pathlib`:

```python
path = pathlib.Path("foo/bar.nc")
basename = path.stem + path.suffix # bar.nc
```

jhkennedy replied:
@agoodm you can do it with a single line:

```python
name = Path('foo/bar.nc').name
```

Dumping strings to file is also a single line:
```python
out_path = Path.cwd() / name
out_path.with_suffix('.refs.json').write_text(json.dumps(refs, indent=2))
out_path.with_suffix('.stac.json').write_text(json.dumps(stac, indent=2))
```
notably, this strips the current extension and replaces it with the provided one.

And for parsing:
```python
stac = json.loads(Path('bar.stac.json').read_text())
```

agoodm replied:
Fair enough, the 2nd example in particular looks like a compelling enough reason to switch.

agoodm commented on src/kerchunk/gran_to_cat.py:
```diff
@@ -0,0 +1,197 @@
+import os
+import warnings
+import fsspec
+import ujson
```
i have had issues with parsing attirbutes from kerchunk references using `orjson`. Since we really don't need the extra speed anyways for this particular use-case I'll switch to using the stdlib `json` module.

agoodm commented:
@jhkennedy I went and addressed most of your comments but regarding your issue I could use some additional information about your package versions (mainly kerchunk, h5py, and netCDF4) since after these changes my code still works in my environment.

jhkennedy commented:
@agoodm I included the listing of the environment in my comment (`env_list.txt`) which has the versions for every package in the environment.

agoodm commented:
@jhkennedy Sorry for missing that!

I was unable to track down which package is causing the inconsistency but I realized that my code should be more robust in handling this case anyway, so I made a quick change. Please try running it again with the latest commit.

agoodm commented:
This latest version now uses pathlib with your suggested changes. how does it look to you now?

jhkennedy commented:
> This latest version now uses pathlib with your suggested changes. How does it look to you now?

@agoodm looks good! 

I've finished my review of the stac item (below) and have a couple of questions about how you're using the refs.

---

#### kerchunk refs

@agoodm in the stac item, I don't see the entire refs encoded in the STAC item; are you just using the `chunks`, `shape`, and `attrs` for each variable in `item['properties']['cube:variables']`, or are you expecting to use the refs file (either directly off S3 or separately encoded in the catalog)?

Overall, it'd be good to know exactly what you intend to use from the STAC item.

---

#### stac item

I've reviewed the STAC item from a search and discovery standpoint since that's the primary purpose of the STAC catalog (e.g., facilitating the queries @betolink will need to do). I don't know enough about how you're using the STAC item, so some of these points may affect that.

Using this granule as the test again:
```
s3://its-live-data/velocity_image_pair/landsatOLI/v02/N70W040/LC08_L1GT_004010_20140206_20200912_02_T2_X_LC08_L1GT_004010_20140529_20200911_02_T2_G0120V02_P008.nc
```

I have a few comments/suggestions on the generated STAC item:
1. `id` shouldn't have the `.nc` extension
2. for the `cube:dimensions` property, 
   1. each dimension `reference_system` is a proj json object, but could just be the EPSG code: https://github.com/stac-extensions/datacube?tab=readme-ov-file#horizontal-spatial-raster-dimension-object
   2. there's no `description` for the dimension, but I'm not sure if it is worth putting "x coordinate of projection in meters", so maybe that's fine
3. for the `cube:variables` property, 
   1. the mapping variable is missing, so the `grid_mapping` attribute doesn't really make sense to include in each variable
   2. For these variables, I think we can drop `attrs`, `shapes`, and `chunks` entirely (they aren't in the cube extension spec) -- are they needed for your kerchunk work?
4. for the assets,
   1. we should include the thumbnail and browse image
   2. we should describe them a bit better by including the type, description, and title as well as the href: https://github.com/radiantearth/stac-spec/blob/master/commons/assets.md#asset-object 
   4. I think the asset key should be "data" so it's easy to select that asset for any item
5. we'll need a bunch more metadata,
   1. most everything should be included from the `img_pair_info` variable as item properties
   2. we'll need most of the information encoded in the variable name as well in the item properties
   3. we'll need the ITS_LIVE datacube "tile" in the item properties
   3. we'll need some of the global attributes in the item properties

I've edited the output stac item for how I'd like to see it:
[LC08_L1GT_004010_20140206_20200912_02_T2_X_LC08_L1GT_004010_20140529_20200911_02_T2_G0120V02_P008.stac-joe.json](https://github.com/user-attachments/files/17168246/LC08_L1GT_004010_20140206_20200912_02_T2_X_LC08_L1GT_004010_20140529_20200911_02_T2_G0120V02_P008.stac-joe.json)

which results in this mini catalog using `stac_geoparquet` (zipped b/c GitHub):
[its_live_mini_catalog.zip](https://github.com/user-attachments/files/17168566/its_live_mini_catalog.zip)

But there are a few things there that may need to change. @betolink @alex-s-gardner @markf6  it'd be good for y'all to review my proposed STAC item as well. I've got a few open questions regarding the item:

- [ ] Is the "roles" specifier for assets valuable? It's not required and the spec implies it's mostly for items with multiple data assets. 
- [ ] Are we happy with the item property names? There are some changes as compared to the netCDF metadata. The major one is:
    - [ ] In @betolink's db he uses `sat1` and `sat2`, while in the netCDF granule we use `img1` and `img2`. I've used `reference` and `secondary` as that's more common in the remote sensing/InSAR world (`ref` and `sec` could also work) and is more descriptive, I think.      
- [ ] Is there metadata in the granule I missed?

jhkennedy commented:
@agoodm I'm happy to make some (all?) of these changes myself -- do you want me to just push to your branch or open a PR to your branch, or do you want to handle all of it?

agoodm commented:
> @agoodm I'm happy to make some (all?) of these changes myself -- do you want me to just push to your branch or open a PR to your branch, or do you want to handle all of it?

You can go ahead and apply these yourself, none of them should have any major impact with my planned virtual cubes work. As far as that is concerned the additional STAC metadata is primarily useful for potential ways to narrow down queries when arrow expands these into separate columns in the geoparquet. However we decided to keep the kerchunk JSON entirely separate for now since it is a bit cleaner without it and also easier and faster to decode later this way.
